daffodil-test/src/test/resources/org/apache/daffodil/layers/layers.tdml

```diff
@@ -383,7 +383,7 @@ jones,arya,cat,1986-02-19
 
 We gzip that, and then we must prepend that with the length (as a binary 4-byte int) before we base64 encode.
   -->
-      <tdml:documentPart type="text"  replaceDFDLEntities="true"><![CDATA[AAAAcx+LCAAAAAAAAAAtyUEKgCAQheG94E1mIDWittG+M0xpaNQIo5tuX0Kb98P7LioVjiTf3sn7%CR;%LF;K8CyzlqVO9UIkrcgFTYh9pnBTOOInUPba3XmyOX7WiEGlqfxgJ1B6xpzKEDyEOxUf7JoJq1e/RI4%CR;%LF;wXIAAAA=--END--]]></tdml:documentPart>
+      <tdml:documentPart type="text"  replaceDFDLEntities="true"><![CDATA[AAAAcx+LCAAAAAAAAP8tyUEKgCAQheG94E1mIDWittG+M0xpaNQIo5tuX0Kb98P7LioVjiTf3sn7%CR;%LF;K8CyzlqVO9UIkrcgFTYh9pnBTOOInUPba3XmyOX7WiEGlqfxgJ1B6xpzKEDyEOxUf7JoJq1e/RI4%CR;%LF;wXIAAAA=--END--]]></tdml:documentPart>
```

Comment:
These two tests show the difference between the GZIPOutputStream in Java 16 and older versions of Java, as mentioned in the commit message. With these changes, these tests will only pass on Java 16.

Any ideas on how to best resolve this? Do we just accept this difference and add a TDML Runner feature similar to "implementation" that skips the test it isn't the right Java version?

Reply:
I was thinking one alternative would be to create something like a ``GZIPFixerOutputStream`` that acts as a proxy betwen the GZIPOutputStream and the underlying output stream. This "fixer" could change the 10th byte to 255 and pass all other bytes through unchanged, so regardless of the java version we get consistent gzip output.

Might look something like this:
```scala
class GZIPFixerOutputStream(os: OutputStream) extends OutputStream {
  private val position = 0
  override def write(byte: Byte) {
    if (position == 10) os.write(255) else os.write(byte)
    postition += 1
  }
}

class GZIPTransformer {
  ...
  override protected def wrapLayerEncoder(jos: java.io.OutputStream): java.io.OutputStream = {
    val fixer = new GZIPFixerOutputStream(jos)
    val s = new java.util.zip.GZIPOutputStream(fix)
    s
  }
}
```
It's a bit of a hack, but at least we get consistent behavior regardless of the Java version?

Reply:
I think a layer which fixes the 10th byte's value for consistency seems like a good idea, but I'm not familiar with how layers are added and used.  Would this layer slow down performance a lot because of the single byte writes?  If there is a big performance hit, can the layer's use be limited to only these two tests?  

I don't think we need to change the TDML schema to allow running tests only on certain JDK versions.  We can add an if statement directly to the Scala test runner class methods which runs these tests so it runs either the pre-Java 16 test case or the Java 16 test case.

Reply:
Yeah, writing bytes one at a time is probably a bad idea, and was more just a snippet to get some discussions going. I've added a commit that adds this new GZIP fixer thing with extra logic so that it is essentially just a passthrouh to the underlying OutputStream after bad byte as been fixed. So there first 10 bytes or so might be a little slower, and there's going to be little bit of overhead to proxy calls to the underlying OutputStream, but I suspect that's minimal compared to the work needed to actually compress data.

This does mean the GZIP layer is going to create slightly different data than before. If we really care, we could add a compatability tunable, and this fixer could set that 9th byte to 0 or 255 depending if you want the old or new Java behavior. Not sure if that's worth it though. But that seems like a bit of work for one byte that most people wont' care about. But at least we are now consistent regarldess of Java version, which is probably more important.

Reply:
So this behavior change for Java 16 is permanent, and the old behavior was clearly wrong, so we should adopt this new behavior going forward, and the fixer should be applied to Java releases prior to this change (actually, what revision introduced this fix, it may be earlier than 16?)

So we should change everything to adopt the Java 16 corrected behavior in the tests, and use the fixer for earlier versions of Java prior to 16 conditionally, based on detecting them. 

I agree a TDML runner enhancement doesn't seem needed.

Question: Does just changing tests to twoPass fix the tests? If we don't actually compare the gzipped data, but rather compare the infosets, as a twoPass test does, maybe that is the easiest thing to do.

Reply:
Ah... mistake. A twoPass test won't fix this, because it requires that a onePass test NOT succeed, but a onePass test will succeed on Java16, so that's not a fix.

Reply:
Yes, this change is part of Java 16. I believe it's just considered a bug prior to Java 16. Here's the release notes where the change is mentioned: https://jdk.java.net/16/release-notes#JDK-8244706

> GZIPOutputStream Sets the GZIP OS Header Field to the Correct Default Value (JDK-8244706)
core-libs/java.util.jar
> Prior to JDK 16, GZIPOutputStream set the OS field in the GZIP header to 0 (meaning FAT filesystem), which does not match the default value specified in section 2.3.1.2 of the GZIP file format specification version 4.3 (RFC 1952).
>
> As of JDK 16, the GZIP OS Header Field is set to 255, which is the default value as defined in RFC 1952.

Take a look at the GZIPFixedOutputStream as part of this PR. This does what you suggest and fixes the output so that this byte is always 255, regardless of the java version.